activerecord/lib/active_record/delay_touching/state.rb

```diff
@@ -0,0 +1,71 @@
+module ActiveRecord
+  # = Active Record Delay Touching
+  module DelayTouching
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
+      def initialize
+        @records = Hash.new { SortedSet.new }
+        @already_updated_records = Hash.new { SortedSet.new }
+      end
+
+      def updated(klass, attrs, records)
+        @already_updated_records[[klass, attrs]] += records
+      end
+
+      # Return the records grouped by class and columns that were touched:
+      #
+      #   {
+      #     [Owner, [:updated_at]]               => SortedSet.new([owner1, owner2]),
+      #     [Pet,   [:neutered_at, :updated_at]] => SortedSet.new([pet1]),
+      #     [Pet,   [:updated_at]]               => SortedSet.new([pet2])
+      #   }
+      #
+      # As a side-effect, clears out the list of records.
+      def get_and_clear_records
+        records = @records
+        @records = Hash.new { SortedSet.new }
+        records
+      end
+
+      def more_records?
+        @records.present?
+      end
+
+      def add_record(record, columns)
```

Comment:
I like some parts of the suggested refactor, but I find the `record.timestamp_attributes_for_update_in_model.concat(extra_columns).map(&:to_sym).sort` line to be a bit much to take in all at once.

I'd like to change it to this:

```ruby
      def add_record(record, columns)
        # Include the standard updated_at column and any additional specified columns
        columns += record.timestamp_attributes_for_update_in_model
        columns = columns.map(&:to_sym).sort

        key = [record.class, columns]
        @records[key] += [record] unless @already_updated_records[key].include?(record)
      end
```
> Also it bothers me a bit the value for the key is wrapped in an array [record]. 

Originally I did this:

```ruby
@records[key] << record unless ...
```

However, it doesn't work because of [the way Ruby handles default Hash values](http://stackoverflow.com/questions/16159370/ruby-hash-default-value-behavior). I think it's fine just to leave it as is.